From 459041eb43474efd6121a0da08c961b9d953a79b Mon Sep 17 00:00:00 2001
From: Flopster101 <nahuelgomez329@gmail.com>
Date: Tue, 30 Sep 2025 13:24:46 -0300
Subject: [PATCH] SystemUI: Cache clock height

Prevents the clock from constantly shifting up and down randomly.

Signed-off-by: Flopster101 <nahuelgomez329@gmail.com>
---
 .../systemui/statusbar/policy/Clock.java      | 33 +++++++++++++++++--
 1 file changed, 30 insertions(+), 3 deletions(-)

diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/policy/Clock.java b/packages/SystemUI/src/com/android/systemui/statusbar/policy/Clock.java
index d4f1c0432e65..19272b71e5be 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/policy/Clock.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/policy/Clock.java
@@ -165,6 +165,9 @@ public class Clock extends TextView implements
     private int mCharsAtCurrentWidth = -1;
     private int mCachedWidth = -1;
 
+    // Cache height to prevent micro-adjustments during font metric changes
+    private int mCachedHeight = -1;
+
     /**
      * Color to be set on this {@link TextView}, when wallpaperTextColor is <b>not</b> utilized.
      */
@@ -458,10 +461,21 @@ public class Clock extends TextView implements
         }
 
         int measuredWidth = getMeasuredWidth();
+        int measuredHeight = getMeasuredHeight();
+
+        // Cache height on first measurement
+        if (mCachedHeight == -1) {
+            mCachedHeight = measuredHeight;
+        }
+
         if (mCachedWidth > measuredWidth) {
-            setMeasuredDimension(mCachedWidth, getMeasuredHeight());
+            setMeasuredDimension(mCachedWidth, mCachedHeight);
         } else {
             mCachedWidth = measuredWidth;
+            // Only update cached height if it's significantly different
+            if (Math.abs(mCachedHeight - measuredHeight) > 2) {
+                mCachedHeight = measuredHeight;
+            }
         }
     }
 
@@ -544,16 +558,29 @@ public class Clock extends TextView implements
     private void reloadDimens() {
         // reset mCachedWidth so the new width would be updated properly when next onMeasure
         mCachedWidth = -1;
+        // Reset height cache to allow recalculation on significant changes
+        mCachedHeight = -1;
 
         FontSizeUtils.updateFontSize(this, R.dimen.status_bar_clock_size);
 
         float fontHeight = getPaint().getFontMetricsInt(null);
         setLineHeight(TypedValue.COMPLEX_UNIT_PX, fontHeight);
 
+        // Use cached height if available to prevent micro-adjustments
         ViewGroup.LayoutParams lp = getLayoutParams();
         if (lp != null) {
-            lp.height = (int) Math.ceil(fontHeight);
-            setLayoutParams(lp);
+            int newHeight = (int) Math.ceil(fontHeight);
+
+            // Only update height if it's significantly different or we don't have a cached height
+            if (mCachedHeight == -1 || Math.abs(mCachedHeight - newHeight) > 2) {
+                lp.height = newHeight;
+                mCachedHeight = newHeight;
+                setLayoutParams(lp);
+            } else {
+                // Use cached height to maintain stability
+                lp.height = mCachedHeight;
+                setLayoutParams(lp);
+            }
         }
     }
 
-- 
2.51.0

